tools/autograd/derivatives.yaml

```diff
@@ -606,13 +606,13 @@
   other: zeros_like(other)
 
 - name: lerp.Scalar(Tensor self, Tensor end, Scalar weight) -> Tensor
-  self: grad * (1 - weight.toDouble())
-  end: grad * weight
+  self: "handle_r_to_c(self.scalar_type(), grad * (1 - (weight.isComplex() ? weight.conj().toComplexDouble() :weight.toDouble())))"
+  end: handle_r_to_c(end.scalar_type(), grad * weight.conj())
```

Comment:
same as above

Reply:
Without `handle_r_to_c`, we get the following error,
```
RuntimeError: Expected isFloatingType(grad.scalar_type()) || (input_is_complex == grad_is_complex) to be true, but got false.  (Could this error message be improved?  If so, please report an enhancement request to PyTorch.)
```
on the following tests,

```
======================================================================= short test summary info ========================================================================
FAILED test/test_ops.py::TestGradientsCPU::test_fn_grad_lerp_cpu_float64 - RuntimeError: Expected isFloatingType(grad.scalar_type()) || (input_is_complex == grad_is_...
FAILED test/test_ops.py::TestGradientsCPU::test_fn_gradgrad_lerp_cpu_float64 - RuntimeError: Expected isFloatingType(grad.scalar_type()) || (input_is_complex == grad...
FAILED test/test_ops.py::TestGradientsCPU::test_inplace_grad_lerp_cpu_float64 - RuntimeError: Expected isFloatingType(grad.scalar_type()) || (input_is_complex == gra...
FAILED test/test_ops.py::TestGradientsCPU::test_inplace_gradgrad_lerp_cpu_float64 - RuntimeError: Expected isFloatingType(grad.scalar_type()) || (input_is_complex ==...
FAILED test/test_ops.py::TestGradientsCUDA::test_fn_grad_lerp_cuda_float64 - RuntimeError: Expected isFloatingType(grad.scalar_type()) || (input_is_complex == grad_i...
FAILED test/test_ops.py::TestGradientsCUDA::test_fn_gradgrad_lerp_cuda_float64 - RuntimeError: Expected isFloatingType(grad.scalar_type()) || (input_is_complex == gr...
FAILED test/test_ops.py::TestGradientsCUDA::test_inplace_grad_lerp_cuda_float64 - RuntimeError: Expected isFloatingType(grad.scalar_type()) || (input_is_complex == g...
FAILED test/test_ops.py::TestGradientsCUDA::test_inplace_gradgrad_lerp_cuda_float64 - RuntimeError: Expected isFloatingType(grad.scalar_type()) || (input_is_complex ...
FAILED test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_lerp_cpu_float32 - RuntimeError: Expected isFloatingType(grad.scalar_type()) || (input_is_comple...
FAILED test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_lerp_cpu_float64 - RuntimeError: Expected isFloatingType(grad.scalar_type()) || (input_is_comple...
FAILED test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_lerp_cuda_float16 - RuntimeError: Expected isFloatingType(grad.scalar_type()) || (input_is_comp...
FAILED test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_lerp_cuda_float32 - RuntimeError: Expected isFloatingType(grad.scalar_type()) || (input_is_comp...
FAILED test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_lerp_cuda_float64 - RuntimeError: Expected isFloatingType(grad.scalar_type()) || (input_is_comp...
================================================ 13 failed, 37 passed, 2 skipped, 8194 deselected, 4 warnings in 13.89s ================================================
```

**Note** : We already have case of `complex tensor inputs` and `float weight`.

Reply:
do you know which input causes this error? I am curious because the above mentioned code snippet and this:
```
>>> start=torch.randn(4)
>>> end=torch.randn(4, dtype=torch.cdouble)
>>> torch.lerp(start, end, 2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: expected dtype float for `end` but got dtype c10::complex<double>
```
are the only inputs I'd expect would cause this error, but both are not valid inputs to the function ...

Reply:
With formula as 
```
- name: lerp.Scalar(Tensor self, Tensor end, Scalar weight) -> Tensor
  self: "grad * (1 - (weight.isComplex() ? weight.conj().toComplexDouble() : weight.toDouble()))"
  end: grad * weight.conj()
```

It is failing on
```
import torch
from torch.autograd import gradcheck, gradgradcheck

x = torch.randn(3, dtype=torch.double, requires_grad=True)
y = torch.randn(3, dtype=torch.double)

out = torch.lerp(x, y, 0.4)
out.sum().backward()
```
Which is really odd! Will investigate.

Reply:
I see what is happening! Return type of ternary operator is deduced to be complex and hence even for else branch (double case) it promotes double to complex.

Thanks for point this out!